code/model/DMSDocumentSet.php

```diff
@@ -230,40 +232,65 @@ public function onBeforeWrite()
 
     /**
      * Retrieve a list of the documents in this set. An extension hook is provided before the result is returned.
-     *
-     * @return ArrayList|null
      */
     public function saveLinkedDocuments()
     {
-        // Documents that belong to just this set.
-        /** @var ManyManyList $originals */
-        $originals = $this->Documents();
-        if (!(empty($this->KeyValuePairs)) && $this->isChanged('KeyValuePairs')) {
-            $keyValuesPair = Convert::json2array($this->KeyValuePairs);
-            /** @var DMSDocument $dmsDoc */
-            $dmsDoc = singleton('DMSDocument');
-            $context = $dmsDoc->getDefaultSearchContext();
+        if (empty($this->KeyValuePairs) || !$this->isChanged('KeyValuePairs')) {
+            return;
+        }
 
-            $sortBy = $this->SortBy ? $this->SortBy : 'LastEdited';
-            $sortByDirection = $this->SortByDirection ? $this->SortByDirection : 'DESC';
-            $sortedBy = sprintf('%s %s', $sortBy, $sortByDirection);
-            /** @var DataList $documents */
-            $documents = $context->getResults($keyValuesPair, $sortedBy);
-            $now = SS_Datetime::now()->Rfc2822();
-            $documents = $documents->where(
-                "\"EmbargoedIndefinitely\" = 0 AND ".
-                " \"EmbargoedUntilPublished\" = 0 AND ".
-                "(\"EmbargoedUntilDate\" IS NULL OR " .
-                "(\"EmbargoedUntilDate\" IS NOT NULL AND '{$now}' >= \"EmbargoedUntilDate\")) AND " .
-                "\"ExpireAtDate\" IS NULL OR (\"ExpireAtDate\" IS NOT NULL AND '{$now}' < \"ExpireAtDate\")"
-            );
+        $keyValuesPair = Convert::json2array($this->KeyValuePairs);
 
-            // Remove all BelongsToSet as the rules have changed
-            $originals->removeByFilter('"BelongsToSet" = 0');
-            foreach ($documents as $document) {
-                $originals->add($document, array('BelongsToSet' => 0));
-            }
+        /** @var DMSDocument $dmsDoc */
+        $dmsDoc = singleton('DMSDocument');
+        $context = $dmsDoc->getDefaultSearchContext();
+
+        $sortBy = $this->SortBy ? $this->SortBy : 'LastEdited';
+        $sortByDirection = $this->SortByDirection ? $this->SortByDirection : 'DESC';
+        $sortedBy = sprintf('%s %s', $sortBy, $sortByDirection);
+
+        /** @var DataList $documents */
+        $documents = $context->getResults($keyValuesPair, $sortedBy);
+        $documents = $this->addEmbargoConditions($documents);
+        $documents = $this->addQueryBuilderSearchResults($documents);
```

Comment:
Minor note: We're not using this last $documents. Maybe just call $this-> addQueryBuilderSearchResults($documents).

I like how you cleaned up this method though.